src/tactic/move_add.lean

```diff
@@ -0,0 +1,397 @@
+* next the list of elements of `sl` that came from an element of `lp` whose boolean was `ff`, and
+* finally the ununified elements of `sl`.
+
+The ununified elements of `sl` get used for error management: they keep track of which user inputs
+are superfluous. -/
+meta def move_left_or_right : list (bool × expr) → list expr → list bool →
+  tactic (list expr × list expr × list expr × list bool)
+| [] sl is_unused      := return ([], [], sl, is_unused)
+| (be::l) sl is_unused :=
+  do (ex :: hs) ← sl.mfilter $ λ e', succeeds $ unify be.2 e' |
+    move_left_or_right l sl (is_unused.append [tt]),
+  (l1, l2, l3, is_unused) ← move_left_or_right l (sl.erase ex) (is_unused.append [ff]),
+  if be.1 then return (ex::l1, l2, l3, is_unused) else return (l1, ex::l2, l3, is_unused)
+
+/--  Given a list of pairs `α × pexpr`, we convert it to a list of `α × expr`. -/
+meta def snd_to_expr {α : Type*} (lp : list (α × pexpr)) : tactic (list (α × expr)) :=
+lp.mmap $ λ x : α × pexpr, do
+  e ← to_expr x.2 tt ff,
+  return (x.1, e)
+
+/--  We combine `snd_to_expr` and `move_left_or_right`, and then some:
+1. we convert a list pairs `bool × pexpr` to a list of pairs `bool × expr`,
+2. we use the extra input `sl : list expr` to perform the unification and sorting step
+   `move_left_or_right`,
+3. we jam the third factor inside the first two.
+-/
+meta def final_sort (lp : list (bool × pexpr)) (sl : list expr) : tactic (list expr × list bool) :=
+do
+  lp_exp : list (bool × expr) ← snd_to_expr lp,
+  (l1, l2, l3, is_unused) ← move_left_or_right lp_exp sl [],
+  return (l1 ++ l3 ++ l2, is_unused)
+
+/-- `sorted_sum` takes an optional location name `hyp` for where it will be applied, a list `ll` of
+`bool × pexpr` (arising as the user-provided input to `move_op`) and an expression `e`.
+
+`sorted_sum hyp ll e` returns an ordered sum of the terms of `e`, where the order is
+determined using the `final_sort` applied to `ll` and `e`.
+
+We use this function for expressions in an (additive) commutative semigroup. -/
+meta def sorted_sum (hyp : option name) (ll : list (bool × pexpr)) (f : pexpr) (e : expr) :
+  tactic (list bool) :=
+do
+  f ← to_expr f tt ff,
+  lisu ← tactic.list_binary_operands f e,
+  (sli, is_unused) ← final_sort ll lisu,
+  match sli with
+  | []       := return is_unused
+  | (eₕ::es) := do
+    t ← infer_type e,
+    fe ← to_expr ``(%%f : %%t → %%t → %%t ),
+    e' ← es.mfoldl (λ eₗ eᵣ, mk_app fe.get_app_fn.const_name [eₗ, eᵣ]) eₕ,
+    e_eq ← mk_app `eq [e, e'],
+    e_eq_fmt ← pp e_eq,
+    h ← solve_aux e_eq $
+      reflexivity <|>
+      `[{ simp only [add_comm, add_assoc, add_left_comm], done, }] <|>
+      `[{ simp only [mul_comm, mul_assoc, mul_left_comm], done, }] <|>
+      fail format!"failed to prove:\n\n{e_eq_fmt}",
+    match hyp with
+    | some loc := do
+      try $ get_local loc >>= rewrite_hyp h.2,
+      pure is_unused
+    | none     := do
+      try $ rewrite_target h.2,
```

Comment:
This feels a bit fragile - rather than doing a global rewrite of the expression we found earlier and hoping we match the same one, can we somehow rewrite exactly the occurence that we found like `conv` does? I'm sure there must be some infrastructure for this somewhere...

Reply:
Ok, it is indeed fragile, since the term itself may be changed and no longer be available for the rewrite.  Here is an example.  Trying `move_add a * _`

```lean
import tactic.move_add

example {F : Type*} [add_comm_semigroup F] [has_le F] {a b c : F} : (a + b) ≤ (a + b + c) :=
by move_add a
--  ⊢ (b + a) ≤ (b + a + c)
```
whereas I would probably expect the goal to become `(b + a) ≤ (b + c + a)`, moving all occurrences of `a` to the rightmost available position.  I suspect that there are better examples where, once one rewrite takes place, the other rewrite is actually no longer available.  Thus, using the `try` was my way of having a tactic that at least does something rather than failing, even though it may not be exactly what the user intended!